### PR TITLE
c8d: Fix checking TLS handshake for insecure registries

### PR DESCRIPTION
**Pull request by rumpl**
**- What I did**

Check for the right error type when receiving an error from an unsecure registry

**- How I did it**

By checking the error type...

**- How to verify it**

Try and pull an image from an insecure registry

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/containerd/resolver.go [hunk 1]
```diff
@@ -72,12 +72,15 @@ type httpFallback struct {
 
 func (f httpFallback) RoundTrip(r *http.Request) (*http.Response, error) {
 	resp, err := f.super.RoundTrip(r)
-	if err != nil {
-		if strings.Contains(err.Error(), "http: server gave HTTP response to HTTPS client") {
-			plain := r.Clone(r.Context())
-			plain.URL.Scheme = "http"
-			return http.DefaultTransport.RoundTrip(plain)
-		}
+	if tlsErr, ok := err.(tls.RecordHeaderError); ok && string(tlsErr.RecordHeader[:]) == "HTTP/" {
```
[the remark this answers is not in the document]
Comment (rumpl) [reply daemon/containerd/resolver.go:75]: I don't think there is a risk of the erorr being wrapped but let's use errors.As anyway

**Comment by thaJeztah** [comment]
Hmm.. haven't seen this one before (I think?). Not sure if related though; possible regression from this one? Or just a flaky?

- https://github.com/moby/moby/pull/45301

```

[2023-04-12T11:26:42.606Z] === FAIL: cmd/docker-proxy TestSCTP4Proxy (0.36s)
[2023-04-12T11:26:42.606Z] 2023/04/12 11:24:33 Can't forward traffic to backend sctp/127.0.0.1:35270: transport endpoint is already connected
[2023-04-12T11:26:42.606Z]     network_proxy_test.go:161: EOF
```

**Comment by thaJeztah** [comment]
From our slack discussion; we should also look at the CLI probably (and maybe have a utility for this, but that's "TBD");
https://github.com/docker/cli/blob/90f17f3f9f22ad611e511ff727cfc7d724130a1e/cli/registry/client/client.go#L131

```go
	httpTransport, err := c.getHTTPTransportForRepoEndpoint(ctx, repoEndpoint)
	if err != nil {
		if !strings.Contains(err.Error(), "server gave HTTP response to HTTPS client") {
			return nil, err
```

**Comment by rumpl** [comment]
> Or just a flaky?

It's flaky yes, not related

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
@akerouanton @corhere ptal

**Comment by rumpl** [comment]
Also @vvoland and @laurazard :)

**Review by corhere — CHANGES_REQUESTED** [review]
Something's not adding up for me. How is the error returned from `f.super.RoundTrip` simultaneously a `tls.RecordHeaderError` and also has a `.Error()` string which contains `"http: server gave HTTP response to HTTPS client"`? My understanding is that https://github.com/golang/go/issues/44855 was needed because the `tls.RecordHeaderError` value wasn't being bubbled up to the caller.

**Comment by rumpl** [comment]
> Something's not adding up for me. How is the error returned from `f.super.RoundTrip` simultaneously a `tls.RecordHeaderError` and also has a `.Error()` string which contains `"http: server gave HTTP response to HTTPS client"`? My understanding is that [golang/go#44855](https://github.com/golang/go/issues/44855) was needed because the `tls.RecordHeaderError` value wasn't being bubbled up to the caller.

That's the thing, the error returned doesn't contain "http: server ...", the error is `tls: first record does not look like a TLS handshake`, with `%#v`:
```
tls.RecordHeaderError{Msg:"first record does not look like a TLS handshake", RecordHeader:[5]uint8{0x48, 0x54, 0x54, 0x50, 0x2f}, Conn:(*net.TCPConn)(0xc00079a330)}
```

**Comment by rumpl** [comment]
@corhere Note: I tested this locally with `distribution/distribution` and with a Docker Desktop's kubernetes cluster; its images are distributed behind a proxy so that they are not blocked for setups where image access management is active. And this check works for both of them.

This is also the code [present in my fork](https://github.com/rumpl/moby/blob/c8d/daemon/containerd/resolver.go#L74) that's been working for many people with Docker Desktop for a while now

**Comment by corhere** [comment]
@rumpl then the PR description is misleading. You're fixing a check that was previously broken due to checking for the wrong error string. I had the impression that this PR was just refactoring the code with a directly equivalent construct.

**Comment by rumpl** [comment]
I updated the PR description and the commit

**Review by corhere — APPROVED** [review]
Thanks!

**Pull request merged**